### ISSUE TITLE
Add Eagle3 speculative decoding draft model training (speculators)

### ISSUE THREAD

**Issue by Maxusmusti** [Red-Hat-AI-Innovation-Team/training_hub#86]
## Summary

Add custom Eagle3 draft model training as a new algorithm in Training Hub, enabling users to train speculative decoding draft models that match their fine-tuned verifier models. This uses the [speculators](https://github.com/foundation-model-stack/speculators) project for Eagle3 architecture and vLLM for online hidden state extraction.

**Source**: RHAIRFE-2212 (Approved), RHAISTRAT-1797

## Problem

When deploying fine-tuned LLMs with speculative decoding, off-the-shelf draft models perform poorly because they weren't trained on the target model's distribution. Training a custom Eagle3 draft model requires:
- Running the verifier model to generate ground truth responses
- Extracting hidden states from intermediate layers (hundreds of GBs if saved to disk)
- Custom PyTorch scripts for training
- Manual vLLM and multi-GPU management

### Step 1: `generate_responses()` — Ground Truth Response Generation
- User provides input dataset (prompts)
- Training Hub runs verifier model via vLLM to generate responses
- Outputs training dataset of (prompt, response) pairs as Parquet/JSONL
- Supports TP > 1 for large verifier models (e.g., 70B)
- Runs once, outside the training loop

### Step 2: `eagle3_train()` — Online Training with vLLM in Loop
- Accepts training dataset from Step 1
- **Key architectural insight**: vLLM runs IN the training loop, extracting hidden states in-memory per batch — no disk storage for hidden states
- For each batch:
  1. Feed prompts+responses to verifier model via vLLM
  2. Extract hidden states from intermediate layers in-memory
  3. Stream hidden states to Eagle3 draft model trainer
- Builds optimized token vocabulary mapping from training data distribution
- Outputs deployable Eagle3 draft model artifact (weights + tokenizer + vocab mapping)

### Architecture Pattern

This follows Training Hub's Algorithm-Backend-Registry pattern:

```
eagle3_response_gen → Algorithm → Backend (vLLM batch inference)
eagle3_train        → Algorithm → Backend (vLLM hidden state extraction + speculators training)
```

The online training pattern is architecturally similar to the existing verl GRPO backend (which runs vLLM rollout workers alongside FSDP training workers), but streams hidden state tensors rather than generated token sequences.

### vLLM Hidden State Extraction
- **Critical dependency**: Need a stable API to extract hidden states from intermediate transformer layers during vLLM forward passes
- Options: vLLM internal model hooks, custom forward pass wrapper, or speculators' own extraction mechanism
- The speculators project may already have utilities for this — needs investigation

### GPU Memory Partitioning
- Verifier model (large, e.g., 70B) runs via vLLM with TP across multiple GPUs
- Draft model (small, e.g., 1B) trains on a single GPU or subset
- Combined memory must fit: vLLM KV cache (reduced utilization) + draft model + optimizer states + hidden state batch buffer
- Expose `gpu_memory_utilization` parameter to control vLLM's KV cache allocation during training

### Batch-Based Hidden State Streaming
- Cannot extract all hidden states at once (memory-prohibitive for large models)
- Process in batches: for each batch of training samples, run vLLM forward pass → extract hidden states → train draft model → release hidden states
- Similar to verl's weight sync pattern but for hidden state tensors

### Token Vocabulary Mapping
- Map between verifier's (potentially large) vocabulary and draft model's (smaller) vocabulary
- Build mapping from training data token distribution
- Include mapping in output artifact for deployment

## Implementation Plan

1. **Research phase**: Validate speculators library capabilities, vLLM hidden state extraction API, Eagle3 architecture requirements
2. **`generate_responses()` algorithm**: Batch vLLM inference with TP support (simpler, reuses existing vLLM patterns)
3. **`eagle3_train()` algorithm**: Online training loop with vLLM hidden state extraction (complex, core feature)
4. **Vocabulary mapping**: Token distribution analysis and mapping generation
5. **Testing**: End-to-end with representative models, acceptance rate validation

## Key Parameters

```python
# Step 1
generate_responses(
    verifier_model_path="path/to/70B-model",
    data_path="prompts.jsonl",
    output_path="training_data.parquet",
    tp_size=4,
    max_tokens=2048,
    temperature=0.0,  # deterministic for training data
)

# Step 2
eagle3_train(
    verifier_model_path="path/to/70B-model",
    draft_model_config="eagle3-1B",  # or path to speculators config
    data_path="training_data.parquet",
    ckpt_output_dir="./eagle3-output",
    tp_size=4,  # for verifier
    num_epochs=3,
    batch_size=32,
    learning_rate=1e-4,
    gpu_memory_utilization=0.5,  # lower than default to leave room for training
)
```

## Success Criteria

- Users train custom Eagle3 draft models via managed workflow without custom scripts
- Online training streams hidden states in-memory (no disk storage)
- TP support for large verifier models (70B+) without OOM
- Trained draft models achieve measurably higher acceptance rates vs off-the-shelf
- Produces deployable artifact compatible with vLLM speculative decoding

## Open Questions

1. What is the stable API in vLLM for hidden state extraction from intermediate layers?
2. Does speculators handle hidden state extraction natively, or do we need custom hooks?
3. How should GPU memory be partitioned between verifier (vLLM TP) and draft model trainer?
4. What is the minimum acceptance rate improvement target?
5. Should `generate_responses()` support checkpointing for very large datasets?

## Dependencies

- [speculators](https://github.com/foundation-model-stack/speculators) — Eagle3 model architecture and training
- vLLM — Hidden state extraction and verifier model serving
- PyTorch — Training loop and distributed communication (NCCL)

## References

- RHAIRFE-2212: Original feature request
- RHAISTRAT-1797: Strategy document with full technical approach
- Existing pattern reference: `lora_grpo_verl.py` (verl GRPO backend with vLLM in loop)

**Comment by Maxusmusti** [comment]
## Research Findings: Speculators + vLLM Hidden State Extraction

### Key Discovery: Speculators Already Has Everything We Need

The [speculators](https://github.com/vllm-project/speculators) project (now under vllm-project) already provides a complete Eagle3 training pipeline. Our integration is primarily a **wrapper** around their existing scripts, not building from scratch.

---

### Eagle3 Architecture

The draft model is remarkably small:
- **FC Layer**: `Linear(3 * hidden_size → hidden_size)` — projects concatenated hidden states from 3 verifier layers
- **1 Decoder Layer** (default): Modified first layer accepts `2 * hidden_size` input (token embeddings + hidden states). Uses Llama-style architecture regardless of verifier type.
- **LM Head**: `Linear(hidden_size → draft_vocab_size)`
- **Frozen copies**: Verifier's `embed_tokens`, `lm_head`, and final `norm` weights (loaded selectively, NOT the full verifier)

Hidden states come from 3 intermediate layers at positions `[2, num_layers//2, num_layers-3]` by default (configurable via `eagle_aux_hidden_state_layer_ids`).

Loss: KL divergence between draft logits and verifier logits.

### vLLM Hidden State Extraction (v0.18.0+)

vLLM natively supports hidden state extraction via the speculative decoding infrastructure ([PR #33736](https://github.com/vllm-project/vllm/pull/33736)):

```bash
vllm serve MODEL --speculative_config '{
  "method": "extract_hidden_states",
  "num_speculative_tokens": 1,
  "draft_model_config": {
    "hf_config": {"eagle_aux_hidden_state_layer_ids": [2, 16, 29]}
  }
}' --kv_transfer_config '{
  "kv_connector": "ExampleHiddenStatesConnector",
  "kv_role": "kv_producer",
  "kv_connector_extra_config": {"shared_storage_path": "/tmp/hidden_states"}
}'
```

This configures vLLM to capture activations from specified layers during prefill and save as safetensors files. Communication with training uses the standard OpenAI-compatible API — **no direct vLLM code dependency** in speculators.

### Speculators Training Pipeline (Already Implemented)

**Step 1: Data Prep** (`scripts/prepare_data.py`)
- Accepts HuggingFace datasets (sharegpt, ultrachat, custom)
- Applies chat template + tokenization
- Computes token frequency stats for vocabulary mapping
- Output: Arrow dataset on disk

**Step 2: Hidden State Extraction** (offline or online)
- **Offline**: `scripts/data_generation_offline.py` — extract all hidden states to disk as safetensors
- **Online**: `--on-missing generate` flag in training script — requests hidden states from live vLLM on-demand

**Step 3: Training** (`scripts/train.py`)
- Creates draft model from verifier config
- Trains with AdamW, linear/cosine LR schedule
- FSDP distributed training via `torchrun`
- Saves checkpoints compatible with vLLM speculative decoding

**Example timing (2x H100, 5K samples, Qwen3-8B)**: ~9.5 min total (data prep: 16s, vLLM startup: 60s, hidden states: 143s, training 5 epochs: 327s)

### Token Vocabulary Mapping (Built-in)

Speculators handles verifier→draft vocabulary mapping:
- `token_freq.pt` computed during data prep
- `build_vocab_mappings_from_distribution()` selects top-K most frequent tokens
- Produces `t2d` (target-to-draft) and `d2t` (draft-to-target) mapping tensors
- Standalone script: `scripts/build_vocab_mapping.py`

### Model Output Format

Trained models deploy directly to vLLM:
```bash
vllm serve path/to/eagle3-speculator
```

Output contains `config.json` with `speculators_config` (algorithm, verifier reference, proposal methods), `model.safetensors`, and generation config. vLLM auto-detects and configures the speculative decoding pipeline.

---

## Training Modes: Three Approaches

### 1. Offline (Separate Phases)
```
Phase 1: vLLM extracts hidden states → disk (safetensors)
Phase 2: Kill vLLM, train on same GPUs using saved data
```
- **Pros**: Simpler, reuses GPUs, proven pipeline
- **Cons**: Disk I/O, storage (~268MB per 8K sequence for Qwen3-8B with 4 layers)
- **Best for**: Smaller datasets, single-node setups

### 2. Online Separated (Dedicated GPU Pools)
```
GPUs 0-1: vLLM serves verifier model (persistent)
GPUs 2-3: Training runs, requests hidden states on-demand
```
- **Pros**: No disk storage, continuous pipeline
- **Cons**: Requires 2x GPU pools, more complex orchestration
- **Best for**: Large-scale production, multi-node setups

### 3. Online Co-located (Same GPUs, Context Switching)
```
Same GPUs alternate: vLLM inference → sleep → train → sleep → vLLM → ...
```
- **Pros**: Minimal GPU count, no disk storage
- **Cons**: Sleep/wake overhead, memory pressure, similar to ART's time-sharing pattern
- **Risks**: We saw instability with this pattern for Gemma4 ART (vLLM segfaults during LoRA reload). However, Eagle3 training doesn't reload LoRA — it only extracts hidden states, so the failure mode is different.
- **Implementation**: Could follow verl's sleep/wake pattern or ART's GPU time-sharing
- **Best for**: GPU-constrained environments

### Recommendation

Support all three modes via a `mode` parameter:
```python
eagle3_train(
    ...
    mode="offline",       # Phase 1: extract to disk, Phase 2: train (default, simplest)
    mode="online",        # Dedicated GPU pools for vLLM and training
    mode="colocated",     # Same GPUs, context switching
)
```

Start with **offline** (proven, ~10 min for 5K samples), add **online separated** next, then **colocated** as optimization.

---

## Memory Considerations

### Hidden state sizes (FP16, 4 layers extracted):

| Model | Per Sequence (8K tokens) | Batch of 32 | Notes |
|-------|------------------------|-------------|-------|
| Qwen3-8B (4096 hidden) | 268 MB | 8.4 GB | Feasible in-memory |
| Llama 70B (8192 hidden) | 512 MB | 16 GB | Needs batching |

For online modes, hidden states are generated per-batch and discarded after training — memory is bounded by batch size, not dataset size.

### GPU partitioning for online separated mode:

For a 70B verifier on 8 GPUs:
- GPUs 0-3: vLLM with TP=4 (verifier model)
- GPUs 4-5: Draft model training (FSDP, though draft is small enough for 1 GPU)
- GPU memory: vLLM uses `gpu_memory_utilization=0.5` (lower than default) to leave headroom

---

## Proposed Training Hub API

```python
from training_hub import eagle3_train, generate_responses

# Step 1: Generate ground truth responses
generate_responses(
    verifier_model="meta-llama/Llama-3.1-70B-Instruct",
    data_path="prompts.jsonl",
    output_path="./training_data",
    tp_size=4,
    max_tokens=2048,
    temperature=0.0,
)

# Step 2: Train Eagle3 draft model
eagle3_train(
    verifier_model="meta-llama/Llama-3.1-70B-Instruct",
    data_path="./training_data",
    ckpt_output_dir="./eagle3-output",
    draft_vocab_size=32000,
    num_epochs=5,
    learning_rate=1e-4,
    tp_size=4,
    mode="offline",  # "offline" | "online" | "colocated"
    n_train_gpus=2,  # for online/colocated modes
)
```

Under the hood, this orchestrates speculators' scripts:
`prepare_data.py` → `launch_vllm.py` → `data_generation_offline.py` → `train.py`

---

## Dependencies

From speculators' `pyproject.toml`:
- `torch>=2.9.0,<=2.12.0`
- `transformers>=4.56.1,<5.10.0`
- `datasets>=4.0.0`
- `openai>=2.0.0` (for vLLM API client)
- `safetensors`, `numpy>=2.0.0`

**No direct vLLM dependency** — vLLM is used as an external API service. Clean separation.

---

## Next Steps

1. Install speculators and run their example pipeline end-to-end on our H100 nodes
2. Write the Training Hub Algorithm + Backend wrapper
3. Test with a real verifier model (e.g., Qwen3-8B or one of our fine-tuned models)
4. Add online separated mode
5. Investigate co-located mode feasibility

## References

- [Extracting hidden states from vLLM (blog)](https://vllm.ai/blog/2026-03-30-extract-hidden-states)
- [Speculators docs](https://docs.vllm.ai/projects/speculators/en/latest/)
- [vLLM EAGLE 3.1 blog](https://vllm.ai/blog/2026-05-26-eagle-3-1)
- [Speculators online training RFC #335](https://github.com/vllm-project/speculators/issues/335)
- [vLLM hidden state extraction PR #33736](https://github.com/vllm-project/vllm/pull/33736)
- [vLLM Observation Plugin RFC #36998](https://github.com/vllm-project/vllm/issues/36998)
- Example scripts: `examples/train/eagle3_llama3_8b_ultrachat_offline_5k.sh`, `eagle3_qwen3_8b_sharegpt_online_5k.sh`
